linux/tests/threadpool_linux_int/threadpool_linux_int.c

```diff
@@ -0,0 +1,208 @@
+    ASSERT_IS_NOT_NULL(threadpool);
+
+    // Create 1 thread pool
+    LogInfo("Scheduling work item");
+    ASSERT_ARE_EQUAL(int, 0, threadpool_schedule_work(threadpool, threadpool_task_wait_20_millisec, (void*)&thread_counter));
+
+    // assert
+    do
+    {
+        wait_on_address(&thread_counter, 1, UINT32_MAX);
+    } while (thread_counter != 1);
+
+    ASSERT_ARE_EQUAL(int32_t, thread_counter, 1, "Thread counter has timed out");
+
+    // cleanup
+    threadpool_destroy(threadpool);
+    execution_engine_dec_ref(execution_engine);
+}
+
+#define N_WORK_ITEMS 10
+
+TEST_FUNCTION(MU_C3(scheduling_, N_WORK_ITEMS, _work_items))
+{
+    // assert
+    EXECUTION_ENGINE_HANDLE execution_engine = execution_engine_create(NULL);
+    uint32_t num_threads = N_WORK_ITEMS;
+    volatile_atomic int32_t thread_counter = 0;
+
+    THREADPOOL_HANDLE threadpool = threadpool_create(execution_engine);
+    ASSERT_IS_NOT_NULL(threadpool);
+
+    // Create double the amount of threads that is the max
+    LogInfo("Scheduling %" PRIu32 " work item", num_threads);
+    for (uint32_t index = 0; index < num_threads; index++)
+    {
+        ASSERT_ARE_EQUAL(int, 0, threadpool_schedule_work(threadpool, threadpool_task_wait_20_millisec, (void*)&thread_counter));
+    }
+
+    // assert
+    do
+    {
+        wait_on_address(&thread_counter, 1, UINT32_MAX);
+    } while (thread_counter != num_threads);
+
+    ASSERT_ARE_EQUAL(int32_t, thread_counter, num_threads, "Thread counter has timed out");
+
+    // cleanup
+    threadpool_destroy(threadpool);
+    execution_engine_dec_ref(execution_engine);
+}
+
+TEST_FUNCTION(MU_C3(scheduling_, N_WORK_ITEMS, _work_items_with_pool_threads))
+{
+    // assert
+    EXECUTION_ENGINE_PARAMETERS_LINUX params;
+    params.min_thread_count = 1;
+    params.max_thread_count = 16;
+
+    EXECUTION_ENGINE_HANDLE execution_engine = execution_engine_create(&params);
+    uint32_t num_threads = N_WORK_ITEMS;
+    volatile_atomic int32_t thread_counter = 0;
+
+    THREADPOOL_HANDLE threadpool = threadpool_create(execution_engine);
+    ASSERT_IS_NOT_NULL(threadpool);
+
+    // Create double the amount of threads that is the max
+    LogInfo("Scheduling %" PRIu32 " work item", num_threads);
+    for (uint32_t index = 0; index < num_threads; index++)
+    {
+        ASSERT_ARE_EQUAL(int, 0, threadpool_schedule_work(threadpool, threadpool_task_wait_20_millisec, (void*)&thread_counter));
+    }
+
+    do
+    {
+        wait_on_address(&thread_counter, 1, UINT32_MAX);
+    } while (thread_counter != num_threads);
+
+    ASSERT_ARE_EQUAL(int32_t, thread_counter, num_threads, "Thread counter has timed out");
+
+    // cleanup
+    threadpool_destroy(threadpool);
+    execution_engine_dec_ref(execution_engine);
+}
+
+#define N_CHAOS_WORK_ITEMS 100
+
+TEST_FUNCTION(threadpool_chaos_knight)
+{
+    // assert
+    EXECUTION_ENGINE_PARAMETERS_LINUX params;
+    params.min_thread_count = 1;
+    params.max_thread_count = 16;
```

Comment:
I think these tests should be fairly similar between win and linux. Shouldn't we just have them in one translation unit?
Probably the setup has to look a bit different (similar to how we did the tls and tcp ones in the EBS client

Reply:
The Windows threadpool exposes a some windows specific stuff.  We'd end up having a lot of #ifdef WIN32 in it.  The Parameter have a WIN and LINUX suffix.  It would require a bit of work to combine them.  I can create a task to see about combining them.